packages/mui-base/src/unstable_useNumberInput/numberInputReducer.ts

```diff
@@ -0,0 +1,154 @@
+import {
+  NumberInputActionContext,
+  NumberInputReducerAction,
+  NumberInputState,
+  StepDirection,
+} from './useNumberInput.types';
+import { NumberInputActionTypes } from './numberInputAction.types';
+import { clamp, isNumber } from './utils';
+
+// extracted from handleValueChange
+function getClampedValues(rawValue: number | undefined, context: NumberInputActionContext) {
+  const { min, max, step } = context;
+
+  const clampedValue = rawValue === undefined ? '' : clamp(rawValue, min, max, step);
+
+  const newInputValue = clampedValue === undefined ? '' : String(clampedValue);
+
+  return {
+    value: clampedValue,
+    inputValue: newInputValue,
+  };
+}
+
+function stepValue(
+  state: NumberInputState,
+  context: NumberInputActionContext,
+  direction: StepDirection,
+  multiplier: number,
+) {
+  const { value } = state;
+  const { step = 1, min, max } = context;
+
+  if (isNumber(value)) {
+    return {
+      up: value + (step ?? 1) * multiplier,
+      down: value - (step ?? 1) * multiplier,
+    }[direction];
+  }
+
+  return {
+    up: min ?? 0,
+    down: max ?? 0,
+  }[direction];
+}
+
+function handleClamp<State extends NumberInputState>(
+  state: State,
+  context: NumberInputActionContext,
+  inputValue: string,
+) {
+  const { getInputValueAsString } = context;
+
+  const parsedValue = getInputValueAsString(inputValue);
+
+  const intermediateValue =
+    parsedValue === '' || parsedValue === '-' ? undefined : parseInt(parsedValue, 10);
+
+  const clampedValues = getClampedValues(intermediateValue, context);
+
+  return {
+    ...state,
+    ...clampedValues,
+  };
+}
+
+function handleInputChange<State extends NumberInputState>(
+  state: State,
+  context: NumberInputActionContext,
+  inputValue: string,
+) {
+  const { getInputValueAsString } = context;
+
+  const parsedValue = getInputValueAsString(inputValue);
```

Comment:
Nit: technically, a "parsed" value would be a number, while here it's just a string representing the value.
BTW, do you intend developers to provide their own implementation of `getInputValueAsString`? What would be the use case?

Reply:
Ah I haven't fully looked into the terminology yet (including `getInputValueAsString`), so let me rename `parsedValue` to `numberValueAsString` for now to be 100% literal 😄 

The use case would be if a user wanted to accept something like `$50` in the input, we would expose `getInputValueAsString` as a prop (with a better name) to accept a custom function to remove the `$`.

It would also require something like `formatNewValueToDisplayValue()` to add back the `$` when the value is updated.

Reply:
Do you think we need such a two-step process? Wouldn't it be enough if developers could override just the parsing function (that gets whatever string there is in the input and returns a number)?